# HG changeset patch
# User Maxim Dounin <mdounin@mdounin.ru>
# Date 1315324516 -14400
# Node ID 1c8c48040004bee990fc2dd984d27e49ca80b017
# Parent  e854e5abda69f3bdee0f87425f9167dc8cd6adca
Core: protection from cycles with named locations and post_action.
Now redirects to named locations are counted against normal uri changes
limit, and post_action respect this limit as well.  As a result at least the
following (bad) configurations no longer trigger infinite cycles:
1. Post action which recursively triggers post action:
    location / {
        post_action /index.html;
    }
2. Post action pointing to nonexistent named location:
    location / {
        post_action @nonexistent;
    }
3. Recursive error page for 500 (Internal Server Error) pointing to
a nonexistent named location:
    location / {
        recursive_error_pages on;
        error_page 500 @nonexistent;
        return 500;
    }
diff --git a/src/http/ngx_http_core_module.c b/src/http/ngx_http_core_module.c
--- a/src/http/ngx_http_core_module.c
+++ b/src/http/ngx_http_core_module.c
@@ -2525,6 +2525,16 @@ ngx_http_named_location(ngx_http_request
     ngx_http_core_main_conf_t   *cmcf;
 
     r->main->count++;
+    r->uri_changes--;
+
+    if (r->uri_changes == 0) {
+        ngx_log_error(NGX_LOG_ERR, r->connection->log, 0,
+                      "rewrite or internal redirection cycle "
+                      "while redirect to named location \"%V\"", name);
+
+        ngx_http_finalize_request(r, NGX_HTTP_INTERNAL_SERVER_ERROR);
+        return NGX_DONE;
+    }
 
     cscf = ngx_http_get_module_srv_conf(r, ngx_http_core_module);
 
diff --git a/src/http/ngx_http_request.c b/src/http/ngx_http_request.c
--- a/src/http/ngx_http_request.c
+++ b/src/http/ngx_http_request.c
@@ -2896,6 +2896,10 @@ ngx_http_post_action(ngx_http_request_t 
         return NGX_DECLINED;
     }
 
+    if (r->post_action && r->uri_changes == 0) {
+        return NGX_DECLINED;
+    }
+
     ngx_log_debug1(NGX_LOG_DEBUG_HTTP, r->connection->log, 0,
                    "post action: \"%V\"", &clcf->post_action);
 
